extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java

```diff
@@ -0,0 +1,227 @@
+
+    /**
+     * Mapper for accessing connection groups.
+     */
+    @Inject
+    private ConnectionGroupMapper connectionGroupMapper;
+
+    /**
+     * Mapper for manipulating connection group permissions.
+     */
+    @Inject
+    private ConnectionGroupPermissionMapper connectionGroupPermissionMapper;
+    
+    /**
+     * Provider for creating connection groups.
+     */
+    @Inject
+    private Provider<ModeledConnectionGroup> connectionGroupProvider;
+
+    /**
+     * Service for creating and tracking sockets.
+     */
+    @Inject
+    private GuacamoleSocketService socketService;
+    
+    @Override
+    protected DirectoryObjectMapper<ConnectionGroupModel> getObjectMapper() {
+        return connectionGroupMapper;
+    }
+
+    @Override
+    protected ObjectPermissionMapper getPermissionMapper() {
+        return connectionGroupPermissionMapper;
+    }
+
+    @Override
+    protected ModeledConnectionGroup getObjectInstance(AuthenticatedUser currentUser,
+            ConnectionGroupModel model) {
+        ModeledConnectionGroup connectionGroup = connectionGroupProvider.get();
+        connectionGroup.init(currentUser, model);
+        return connectionGroup;
+    }
+
+    @Override
+    protected ConnectionGroupModel getModelInstance(AuthenticatedUser currentUser,
+            final ConnectionGroup object) {
+
+        // Create new ModeledConnectionGroup backed by blank model
+        ConnectionGroupModel model = new ConnectionGroupModel();
+        ModeledConnectionGroup connectionGroup = getObjectInstance(currentUser, model);
+
+        // Set model contents through ModeledConnectionGroup, copying the provided connection group
+        connectionGroup.setParentIdentifier(object.getParentIdentifier());
+        connectionGroup.setName(object.getName());
+        connectionGroup.setType(object.getType());
+
+        return model;
+        
+    }
+
+    @Override
+    protected boolean hasCreatePermission(AuthenticatedUser user)
+            throws GuacamoleException {
+
+        // Return whether user has explicit connection group creation permission
+        SystemPermissionSet permissionSet = user.getUser().getSystemPermissions();
+        return permissionSet.hasPermission(SystemPermission.Type.CREATE_CONNECTION_GROUP);
+
+    }
+
+    @Override
+    protected ObjectPermissionSet getPermissionSet(AuthenticatedUser user)
+            throws GuacamoleException {
+
+        // Return permissions related to connection groups 
+        return user.getUser().getConnectionGroupPermissions();
+
+    }
+
+    @Override
+    protected void validateNewObject(AuthenticatedUser user, ConnectionGroup object)
+            throws GuacamoleException {
+
+        // Name must not be blank
+        if (object.getName().trim().isEmpty())
+            throw new GuacamoleClientException("Connection group names must not be blank.");
+        
+        // FIXME: Do not attempt to create duplicate connection groups
```

Comment:
Nothing but the desire to keep the review from getting huge beyond belief (not that it isn't already).